Introduction¶

In this project I will convert addresses into their equivalent latitude and longitude values. Also, I will use the Foursquare API to explore neighborhoods in New York City. I will use the explore function to get the top 10 venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters by using k-means clustering algorithm. Finally, I will use the Folium library to visualize the neighborhoods in New York City and their emerging clusters.


# Download All Dependent Libraries

In [7]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


# Load and Explore Data Set

The link to the dataset: https://geo.nyu.edu/catalog/nyu_2451_34572

In [8]:
!wget -q -O 'newyork_data.json' https://ibm.box.com/shared/static/fbpwbovar7lf8p5sgddm06cgipa2rxpe.json
print('Data downloaded!')

Data downloaded!


In [9]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [10]:
#Feature give required Data. Hence Exploring and trnasforming to Pandas DF
newyork_data['features'][0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [11]:
neighborhoods_data = newyork_data['features']
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [12]:
#Fill the Dataframe from the resultset
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [13]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [14]:
print('The dataframe has {} boroughs and {} size.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape
    )
)

The dataframe has 5 boroughs and (306, 4) size.


In [15]:
#Get coordinates for NYC

address = 'New York City, NY'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geograpical coordinate of New York City are 40.7308619, -73.9871558.


In [16]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

As We have seen Newyork data comprises of 5 Boroughs. Assume a person wants to plan a 2 days trip to NYC to get the most exposure of the city.Hence we will now explore the neighbourhoods in NYC .

# Function to Explore Neighbourhoods in NeW York

In [17]:
CLIENT_ID = 'LQTUYSMBD0ZFOW5LSENYAG13HWAD5D32OCJYFKCLDA41OHJL' # your Foursquare ID
CLIENT_SECRET = 'Q3WM50KUQBHJQDKZ3MOET3XPQL0BMI5EFXQC2GCOMPVXRDOP' # your Foursquare Secret
VERSION = '20181018' # Foursquare API version
LIMIT = 100
radius = 250
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LQTUYSMBD0ZFOW5LSENYAG13HWAD5D32OCJYFKCLDA41OHJL
CLIENT_SECRET:Q3WM50KUQBHJQDKZ3MOET3XPQL0BMI5EFXQC2GCOMPVXRDOP


In [18]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [19]:
newyork_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [118]:
newyork_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896521       -73.844680   
2  Cooler Runnings Jamaican Restaurant Inc       40.898283       -73.850478   
3                         Carvel Ice Cream       40.890487       -73.848568   
4                            Dunkin Donuts       40.890631       -73.849027   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2  Caribbean Restaurant  
3        Ice Cream Shop  
4            Donut Shop

In [133]:
Final = newyork_venues.filter(['Neighborhood','Neighborhood Latitude','Neighborhood Longitude'], axis =1)
Final = Final.drop_duplicates(subset=None, keep='first', inplace=False)
Final.shape

(306, 3)

In [145]:
reference = newyork_venues.groupby('Neighborhood').max()
reference.shape

(302, 6)

In [22]:
print('There are {} uniques categories.'.format(len(newyork_venues['Venue Category'].unique())))

There are 424 uniques categories.


# Get Top 5 Categories for Each Neighbourhood

In [115]:
# one hot encoding
newyork_onehot = pd.get_dummies(newyork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
newyork_onehot['Neighborhood'] = newyork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [newyork_onehot.columns[-1]] + list(newyork_onehot.columns[:-1])
newyork_onehot = newyork_onehot[fixed_columns]

newyork_onehot.head()


Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  Airport Terminal  Airport Tram  American Restaurant  \
0                   0                 0             0                    0   
1                   0                 0             0                    0   
2                   0                 0             0                    0   
3                   0                 0             0                    0   
4                   0                 0             0                    0   

   Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
0               0             0       0                 0   
1               0             0       0                 0   
2               0             0       0                 0   
3               0             0       0                 0   
4               0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  Auditorium  \
0                     0                 0                   0           0   
1                     0                 0                   0           0   
2                     0                 0                   0           0   
3                     0                 0                   0           0   
4                     0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  Automotive Shop  \
0                      0                    0              0                0   
1                      0                    0              0                0   
2                      0                    0              0                0   
3                      0                    0              0                0   
4                      0                    0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Bath House  Beach  Beach Bar  \
0                 0                 0           0      0          0   
1                 0                 0           0      0          0   
2                 0                 0           0      0          0   
3                 0                 0           0      0          0   
4                 0                 0           0      0          0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  \
0                0         0            0           0                   0   
1                0         0            0           0                   0   
2                0         0            0           0                   0   
3                0         0            0           0                   0   
4                0         0            0           0                 

In [25]:
newyork_onehot.shape

(10272, 424)

In [27]:
newyork_grouped = newyork_onehot.groupby('Neighborhood').mean().reset_index()
newyork_grouped.head()

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0       Allerton          0.0                0.0             0.0   
1       Annadale          0.0                0.0             0.0   
2  Arden Heights          0.0                0.0             0.0   
3      Arlington          0.0                0.0             0.0   
4       Arrochar          0.0                0.0             0.0   

   Afghan Restaurant  African Restaurant  Airport Terminal  Airport Tram  \
0                0.0                 0.0               0.0           0.0   
1                0.0                 0.0               0.0           0.0   
2                0.0                 0.0               0.0           0.0   
3                0.0                 0.0               0.0           0.0   
4                0.0                 0.0               0.0           0.0   

   American Restaurant  Animal Shelter  Antique Shop  Arcade  \
0                 0.00             0.0           0.0     0.0   
1                 0.00             0.0           0.0     0.0   
2                 0.00             0.0           0.0     0.0   
3                 0.25             0.0           0.0     0.0   
4                 0.00             0.0           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0               0.0                     0.0          0.0         0.0   
1               0.0                     0.0          0.0         0.0   
2               0.0                     0.0          0.0         0.0   
3               0.0                     0.0          0.0         0.0   
4               0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0               0.0   
1                  0.0                   0.0               0.0   
2                  0.0                   0.0               0.0   
3                  0.0                   0.0               0.0   
4                  0.0                   0.0               0.0   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0                 0.0         0.0                    0.0                  0.0   
1                 0.0         0.0                    0.0                  0.0   
2                 0.0         0.0                    0.0                  0.0   
3                 0.0         0.0                    0.0                  0.0   
4                 0.0         0.0                    0.0                  0.0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0.0              0.0        0.0         0.0    0.000000     0.0   
1            0.0              0.0        0.0         0.0    0.000000     0.0   
2            0.0              0.0        0.0         0.0    0.000000     0.0   
3            0.0              0.0        0.0         0.0    0.000000     0.0   
4            0.0              0.0        0.0         0.0    0.055556     0.0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  Bath House  \
0   0.0  0.0             0.0               0.0               0.0         0.0   
1   0.0  0.0             0.0               0.0               0.0         0.0   
2   0.0  0.0             0.0               0.0               0.0         0.0   
3   0.0  0.0             0.0               0.0               0.0         0.0   
4   0.0  0.0             0.0               0.0               0.0         0.0   

   Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0    0.0        0.0              0.0       0.0          0.0         0.0   
1    0.0        0.0              0.0       0.0          0.0         0.0   
2    0.0        0.0              0.0       0.0          0.0         0.0   
3    0.0        0.0              0.0       0.0          0.0         0.0   
4    0.0        0.0              0.0       0.0          0.0         0.0   

   Belgian Restaurant  

In [100]:
newyork_grouped.Neighborhood

0                       Allerton
1                       Annadale
2                  Arden Heights
3                      Arlington
4                       Arrochar
5                        Arverne
6                        Astoria
7                Astoria Heights
8                     Auburndale
9                     Bath Beach
10             Battery Park City
11                     Bay Ridge
12                   Bay Terrace
13                    Baychester
14                       Bayside
15                     Bayswater
16                  Bedford Park
17            Bedford Stuyvesant
18                    Beechhurst
19                      Bellaire
20                  Belle Harbor
21                     Bellerose
22                       Belmont
23                   Bensonhurst
24                  Bergen Beach
25                    Blissville
26                    Bloomfield
27                   Boerum Hill
28                  Borough Park
29                  Breezy Point
30        

In [107]:
Temp= neighborhoods.'Neighborhood in newyork_grouped.Neighborhood

#neighborhoods.merge(newyork_grouped, on='Neighborhood',copy=False)
#neighborhoods_final = neighborhoods.merge(newyork_grouped,how='inner', on=['Neighborhood'],indicator=True)
#neighborhoods_final
#neighborhoods_final['_merge'].unique()
#neighborhoods_final = neighborhoods_final[neighborhoods_final['_merge']=='both'].reset_index(drop=True)
#neighborhoods_final.shape

SyntaxError: EOL while scanning string literal (<ipython-input-107-a839ba3b9c85>, line 1)

We have 302 Neighbourhood Groups with category Values. Lets get 5 most common values of each Neighbourhood

In [67]:
num_top_venues = 5

for hood in newyork_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = newyork_grouped[newyork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Allerton----
                venue  freq
0         Supermarket  0.12
1       Deli / Bodega  0.12
2  Chinese Restaurant  0.09
3         Pizza Place  0.09
4                 Spa  0.06


----Annadale----
            venue  freq
0     Pizza Place  0.11
1  Cosmetics Shop  0.11
2      Sports Bar  0.11
3            Food  0.11
4            Park  0.11


----Arden Heights----
              venue  freq
0       Pizza Place  0.25
1          Pharmacy  0.25
2       Coffee Shop  0.25
3          Bus Stop  0.25
4  Pedestrian Plaza  0.00


----Arlington----
                 venue  freq
0             Bus Stop  0.50
1        Deli / Bodega  0.25
2  American Restaurant  0.25
3          Yoga Studio  0.00
4                 Park  0.00


----Arrochar----
                venue  freq
0         Pizza Place  0.11
1       Deli / Bodega  0.11
2            Bus Stop  0.11
3  Italian Restaurant  0.11
4         Supermarket  0.06


----Arverne----
             venue  freq
0        Surf Spot  0.27
1    Metro Station  0.1

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [184]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = newyork_grouped['Neighborhood']

for ind in np.arange(newyork_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(newyork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood            1st Most Common Venue  \
0                     Allerton                    Deli / Bodega   
1                     Annadale                       Restaurant   
2                Arden Heights                   Cluster Labels   
3                    Arlington                   Cluster Labels   
4                     Arrochar               Italian Restaurant   
5                      Arverne                        Surf Spot   
6                      Astoria        Middle Eastern Restaurant   
7              Astoria Heights               Italian Restaurant   
8                   Auburndale               Italian Restaurant   
9                   Bath Beach                       Donut Shop   
10           Battery Park City                      Coffee Shop   
11                   Bay Ridge                      Pizza Place   
12                 Bay Terrace                   Clothing Store   
13                  Baychester              Fried Chicken Joint   
14                     Bayside                              Bar   
15                   Bayswater                   Cluster Labels   
16                Bedford Park                    Deli / Bodega   
17          Bedford Stuyvesant                      Coffee Shop   
18                  Beechhurst                              Gym   
19                    Bellaire               Chinese Restaurant   
20                Belle Harbor                    Deli / Bodega   
21                   Bellerose                    Deli / Bodega   
22                     Belmont               Italian Restaurant   
23                 Bensonhurst               Chinese Restaurant   
24                Bergen Beach                  Harbor / Marina   
25                  Blissville                            Hotel   
26                  Bloomfield                   Cluster Labels   
27                 Boerum Hill                      Coffee Shop   
28                Borough Park                             Bank   
29                Breezy Point              Monument / Landmark   
30                   Briarwood                    Deli / Bodega   
31              Brighton Beach               Russian Restaurant   
32               Broad Channel                      Pizza Place   
33           Broadway Junction                            Diner   
34                   Bronxdale               Italian Restaurant   
35            Brooklyn Heights                      Yoga Studio   
36                  Brookville                    Deli / Bodega   
37                 Brownsville                             Park   
38                  Bulls Head                         Bus Stop   
39                    Bushwick                              Bar   
40                Butler Manor                   Baseball Field   
41             Cambria Heights             Caribbean Restaurant   
42                    Canarsie               Chinese Restaurant   
43               Carnegie Hill                      Pizza Place   
44             Carroll Gardens               Italian Restaurant   
45                 Castle Hill                      Pizza Place   
46           Castleton Corners                      Pizza Place   
47              Central Harlem               African Restaurant   
48                  Charleston                       Shoe Store   
49                     Chelsea                      Coffee Shop   
50                   Chinatown               Chinese Restaurant   
51                 City Island                  Harbor / Marina   
52                   City Line                       Donut Shop   
53                Civic Center             Gym / Fitness Center   
54           Claremont Village                      Bus Station   
55                Clason Point                   Cluster Labels   
56                     Clifton                    Train Station   
57                     Clinton                          Theater   
58                Clinton Hill                      Pizza Place  

# Use K Means Clustering

In [185]:
# set number of clusters
kclusters = 4

newyork_grouped_clustering = newyork_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(newyork_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 
kmeans.labels_

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 2, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,

In [186]:
reference.head()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                   
Allerton                   40.865788              -73.859319   
Annadale                   40.538114              -74.178549   
Arden Heights              40.549286              -74.185887   
Arlington                  40.635325              -74.165104   
Arrochar                   40.596313              -74.067124   

                                           Venue  Venue Latitude  \
Neighborhood                                                       
Allerton                            White Castle       40.866786   
Annadale                            Square Pizza       40.542079   
Arden Heights  Villa Monte Pizzeria & Restaurant       40.552555   
Arlington                     Value Deli & Grill       40.636983   
Arrochar                      Via Ponte Pizzeria       40.599184   

               Venue Longitude Venue Category  Cluster Labels  
Neighborhood                                                   
Allerton            -73.855236    Supermarket               0  
Annadale            -74.175667  Train Station               0  
Arden Heights       -74.184030    Pizza Place               1  
Arlington           -74.161499  Deli / Bodega               1  
Arrochar            -74.062288    Supermarket               0

In [187]:
newyork_merged = reference

# add clustering labels
newyork_merged['Cluster Labels'] = kmeans.labels_

newyork_merged = newyork_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
newyork_merged = newyork_merged.drop(['Venue Latitude','Venue Longitude','Venue Category', 'Venue'], axis =1)
newyork_merged.head() # check the last columns!

Neighborhood Latitude  Neighborhood Longitude  Cluster Labels  \
Neighborhood                                                                   
Allerton                   40.865788              -73.859319               0   
Annadale                   40.538114              -74.178549               0   
Arden Heights              40.549286              -74.185887               1   
Arlington                  40.635325              -74.165104               1   
Arrochar                   40.596313              -74.067124               0   

              1st Most Common Venue 2nd Most Common Venue  \
Neighborhood                                                
Allerton              Deli / Bodega           Supermarket   
Annadale                 Restaurant         Train Station   
Arden Heights        Cluster Labels              Bus Stop   
Arlington            Cluster Labels              Bus Stop   
Arrochar         Italian Restaurant              Bus Stop   

              3rd Most Common Venue 4th Most Common Venue  \
Neighborhood                                                
Allerton         Chinese Restaurant           Pizza Place   
Annadale                       Park          Liquor Store   
Arden Heights           Pizza Place           Coffee Shop   
Arlington             Deli / Bodega   American Restaurant   
Arrochar              Deli / Bodega           Pizza Place   

                   5th Most Common Venue 6th Most Common Venue  \
Neighborhood                                                     
Allerton                             Spa           Bus Station   
Annadale                           Diner            Sports Bar   
Arden Heights                   Pharmacy            Food Truck   
Arlington                    Flea Market               Factory   
Arrochar       Middle Eastern Restaurant            Bagel Shop   

              7th Most Common Venue 8th Most Common Venue  \
Neighborhood                                                
Allerton                       Food    Spanish Restaurant   
Annadale                Pizza Place                  Food   
Arden Heights            Food Stand            Eye Doctor   
Arlington        Falafel Restaurant                  Farm   
Arrochar               Liquor Store     Polish Restaurant   

              9th Most Common Venue 10th Most Common Venue  
Neighborhood                                                
Allerton                 Bike Trail          Grocery Store  
Annadale             Cosmetics Shop                  Field  
Arden Heights               Factory    Fried Chicken Joint  
Arlington            Farmers Market   Fast Food Restaurant  
Arrochar                      Hotel         Sandwich Place

Finalize the Clusters

In [188]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(newyork_merged['Neighborhood Latitude'], newyork_merged['Neighborhood Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [ ]:
newyork_merged.loc[newyork_merged['Cluster Labels'] == 0, newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]